File: server/test/ansi.test.js
'use strict';
const { test } = require('node:test');
const assert = require('node:assert');
const { rgb565, xterm256, parseLine, stripAnsi } = require('../lib/ansi');

const DEF = 0xFFFF;
const ESC = '\x1b';

test('rgb565 packs r,g,b into 16-bit 565', () => {
  assert.equal(rgb565(255, 255, 255), 0xFFFF);
  assert.equal(rgb565(0, 0, 0), 0x0000);
  assert.equal(rgb565(255, 0, 0), 0xF800);
  assert.equal(rgb565(0, 255, 0), 0x07E0);
  assert.equal(rgb565(0, 0, 255), 0x001F);
});

test('xterm256 maps cube + greyscale indices', () => {
  assert.equal(xterm256(196), rgb565(255, 0, 0));   // cube bright red
  assert.equal(xterm256(231), rgb565(255, 255, 255)); // cube white
  assert.equal(xterm256(16), rgb565(0, 0, 0));      // cube black
  assert.equal(xterm256(255), rgb565(238, 238, 238)); // top greyscale
});

test('stripAnsi removes escape sequences, keeping text', () => {
  assert.equal(stripAnsi(`${ESC}[31mhi${ESC}[0m`), 'hi');
  assert.equal(stripAnsi(`plain`), 'plain');
  assert.equal(stripAnsi(`${ESC}[38;5;246m x ${ESC}[39m`), ' x ');
});

test('parseLine returns clean text + the color at the first visible char', () => {
  const r = parseLine(`${ESC}[38;5;231muser typed this${ESC}[39m`, DEF);
  assert.equal(r.text, 'user typed this');
  assert.equal(r.color, xterm256(231));
});

test('parseLine uses the default color when no color is set', () => {
  const r = parseLine('just text', DEF);
  assert.equal(r.text, 'just text');
  assert.equal(r.color, DEF);
});

test('parseLine honors 24-bit truecolor (38;2;r;g;b)', () => {
  const r = parseLine(`${ESC}[38;2;255;0;0mred`, DEF);
  assert.equal(r.color, rgb565(255, 0, 0));
});

test('parseLine ignores leading spaces when picking the line color', () => {
  const r = parseLine(`   ${ESC}[38;5;196mX`, DEF);
  assert.equal(r.text, '   X');
  assert.equal(r.color, xterm256(196));
});

File: server-go/internal/screen/ansi_test.go
package screen

import "testing"

const esc = "\x1b"

func TestRGB565(t *testing.T) {
	cases := []struct {
		r, g, b int
		want    uint16
	}{
		{255, 255, 255, 0xFFFF}, {0, 0, 0, 0x0000}, {255, 0, 0, 0xF800},
		{0, 255, 0, 0x07E0}, {0, 0, 255, 0x001F},
	}
	for _, c := range cases {
		if got := RGB565(c.r, c.g, c.b); got != c.want {
			t.Fatalf("RGB565(%d,%d,%d)=%#x want %#x", c.r, c.g, c.b, got, c.want)
		}
	}
}

func TestXterm256(t *testing.T) {
	if Xterm256(196) != RGB565(255, 0, 0) {
		t.Fatal("196")
	}
	if Xterm256(231) != RGB565(255, 255, 255) {
		t.Fatal("231")
	}
	if Xterm256(16) != RGB565(0, 0, 0) {
		t.Fatal("16")
	}
	if Xterm256(255) != RGB565(238, 238, 238) {
		t.Fatal("255")
	}
}

func TestStripAnsi(t *testing.T) {
	if StripAnsi(esc+"[31mhi"+esc+"[0m") != "hi" {
		t.Fatal("colored")
	}
	if StripAnsi("plain") != "plain" {
		t.Fatal("plain")
	}
	if StripAnsi(esc+"[38;5;246m x "+esc+"[39m") != " x " {
		t.Fatal("256")
	}
}

func TestParseLine(t *testing.T) {
	txt, col := ParseLine(esc+"[38;5;231muser typed this"+esc+"[39m", 0xFFFF)
	if txt != "user typed this" || col != Xterm256(231) {
		t.Fatalf("got %q %#x", txt, col)
	}
	txt, col = ParseLine("just text", 0xFFFF)
	if txt != "just text" || col != 0xFFFF {
		t.Fatal("default color")
	}
	_, col = ParseLine(esc+"[38;2;255;0;0mred", 0xFFFF)
	if col != RGB565(255, 0, 0) {
		t.Fatal("truecolor")
	}
	txt, col = ParseLine("   "+esc+"[38;5;196mX", 0xFFFF)
	if txt != "   X" || col != Xterm256(196) {
		t.Fatalf("leading spaces got %q %#x", txt, col)
	}
}

func TestParseLineBoldBrightens(t *testing.T) {
	_, col := ParseLine(esc+"[1;31mX", 0xFFFF)
	if col != sys16[9] {
		t.Fatalf("bold+31 got %#x want bright red %#x", col, sys16[9])
	}
	_, col = ParseLine(esc+"[1m"+esc+"[32mX", 0xFFFF)
	if col != sys16[10] {
		t.Fatalf("bold then 32 got %#x want bright green %#x", col, sys16[10])
	}
	_, col = ParseLine(esc+"[31mX", 0xFFFF)
	if col != sys16[1] {
		t.Fatalf("plain 31 got %#x want normal red %#x", col, sys16[1])
	}
}

func TestParseLineBoldBrightensRegardlessOfOrder(t *testing.T) {
	_, col := ParseLine(esc+"[31;1mX", 0xFFFF)
	if col != sys16[9] {
		t.Fatalf("31;1 got %#x want bright red %#x", col, sys16[9])
	}
	_, col = ParseLine(esc+"[34m"+esc+"[1mX", 0xFFFF)
	if col != sys16[12] {
		t.Fatalf("34m then 1m before text got %#x want bright blue %#x", col, sys16[12])
	}
	_, col = ParseLine(esc+"[1;31;22mX", 0xFFFF)
	if col != sys16[1] {
		t.Fatalf("1;31;22 (unbolded) got %#x want normal red %#x", col, sys16[1])
	}
	_, col = ParseLine(esc+"[1;38;5;196mX", 0xFFFF)
	if col != Xterm256(196) {
		t.Fatalf("bold + 256-color got %#x want %#x (256 unaffected by bold)", col, Xterm256(196))
	}
}
